boot/bootstate20.go

```diff
@@ -21,329 +21,511 @@ package boot
-	// 2. Update kernel_status to "try"
-	//
-	// This is because if we get rebooted in between 1 and 2, kernel_status
-	// is still unset and boot scripts proceeds to boot with the old kernel,
-	// effectively ignoring the try-kernel symlink.
-	// If we did it in the opposite order however, we would set kernel_status to
-	// "try" and then get rebooted before we could create the try-kernel
-	// symlink, so the bootloader would try to boot from the non-existent
-	// try-kernel symlink and become broken.
+	return bsmark, nil
+}
 
-	// However, if we have successfully just booted from a try-kernel and are
-	// marking it successful (this implies that snap_kernel=="trying" as set by
-	// the boot script), we need to do the following order (since we have the
-	// added complexity of moving the kernel symlink):
-	// 1. Update kernel_status to ""
-	// 2. Move kernel symlink to point to the new try kernel
-	// 3. Remove try-kernel symlink
-	//
-	// If we got rebooted after step 1, then the bootloader is booting the wrong
-	// kernel, but is at least booting a known good kernel and snapd in
-	// user-space would be able to figure out the inconsistency.
-	// If we got rebooted after step 2, the bootloader would boot from the new
-	// try-kernel which is okay and all that's left is for snapd to cleanup the
-	// left-over try-kernel symlink.
-	// If instead we had moved the kernel symlink first to point to the new try
-	// kernel, and got rebooted before the kernel_status was updated, we would
-	// have kernel_status="trying" which would cause the bootloader to think
-	// the boot failed, and revert to booting using the kernel symlink, but that
-	// now points to the new kernel we were trying and we did not successfully
-	// boot from that kernel to know we should trust it.
-	// The try-kernel symlink removal should happen last because it will not
-	// affect anything, except that if it was removed before updating
-	// kernel_status to "", the bootloader will think that the try kernel failed
-	// to boot and fall back to booting the old kernel which is safe.
-
-	// enable snaps that need to be enabled
-	// * step 1 for about to try an update
-	// * NOP for successful boot after an update
-	if u20.snapToTry != nil {
-		switch u20.typ {
-		case snap.TypeKernel:
-			err := u20.ebl.EnableTryKernel(u20.snapToTry)
-			if err != nil {
-				return err
-			}
-		default:
-			return fmt.Errorf("boot base update not implemented yet")
+// genericMarkSuccessful sets the necessary boot variables, etc. to mark the
+// given boot snap as successful and a valid rollback target. If err is nil,
+// then the first return value is guaranteed to always be non-nil.
+func genericMarkSuccessful(b bootState, update bootStateUpdate) (bsmark *bootState20MarkSuccessful, trySnap snap.PlaceInfo, shouldClean bool, err error) {
+	// either combine the provided bootStateUpdate with a new one for this type
+	// or create a new one for this type
+	var ok bool
+	if update != nil {
```

Comment:
This could probably be combined into `threadBootState20MarkSuccessful()`, so that we have
```
func genericMarkSuccessful(b bootState, update bootStateUpdate) (bsmark *bootState20MarkSuccessful, trySnap snap.PlaceInfo, shouldClean bool, err error) {
	bsmark, err := threadBootState20MarkSuccessful(update)
	if err != nil {
		return nil, nil, false, err
	}
	...
```
```